### PR TITLE
feat(application): merge commands within groups

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox`?

-----

Allows Snapcraft to override default commands like the `InitCommand` and `PackCommand` without a hacky override ([source](https://github.com/canonical/snapcraft/pull/5129/files#r1813199559)).

### Features
- Applications can add a command to a `CommandGroup` where craft-application already provides a default command with the same name.  The application's command will replace the default command. 

### Changes
- `CommandGroup.ordered` is no longer discarded.
- The Lifecycle CommandGroup is ordered.
- If an application registers multiple CommandGroups with the same group name, they will not be merged.
    - I'm OK with this change because it is an odd thing to do, no applications do it, and I would consider it an incorrect way to use craft application.

(CRAFT-3543)

### CODE REVIEW

File: craft_application/application.py [hunk 1]
```diff
@@ -176,22 +177,77 @@ def app_config(self) -> dict[str, Any]:
...
+
+        merged[lifeycle_default_commands.name] = self._merge_defaults(
+            app_commands=merged.get(lifeycle_default_commands.name),
+            default_commands=lifeycle_default_commands,
+        )
+        merged[other_default_commands.name] = self._merge_defaults(
+            app_commands=merged.get(other_default_commands.name),
+            default_commands=other_default_commands,
+        )
+
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply craft_application/application.py:203]: Correct, that is unchanged.  It's a craft-cli limitation but I added a note to the docstring.

File: craft_application/application.py [hunk 2]
```diff
@@ -176,22 +177,77 @@ def app_config(self) -> dict[str, Any]:
...
+        # for lookup of commands by name
+        app_commands_dict = {command.name: command for command in app_commands.commands}
+
+        merged_commands: list[type[craft_cli.BaseCommand]] = []
+        processed_command_names: set[str] = set()
+
+        for default_command in default_commands.commands:
+            # prefer the application command if it exists
+            if default_command.name in app_commands_dict:
+                merged_commands.append(app_commands_dict[default_command.name])
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply craft_application/application.py]: Good idea, I added some trace logs.

[the remark this answers is not in the document]
Comment (mr-cal) [reply craft_application/application.py]: I changed them to debug, I forgot this was in the twilight zone before the dispatcher is created and verbosity is still brief.

**Review by dariuszd21 — APPROVED** [review]
Great!

File: craft_application/application.py [hunk 3]
```diff
@@ -176,22 +177,85 @@ def app_config(self) -> dict[str, Any]:
...
+        )
+
+        return list(merged.values())
+
+    def _merge_defaults(
+        self,
+        *,
+        app_commands: craft_cli.CommandGroup | None,
+        default_commands: craft_cli.CommandGroup,
+    ) -> CommandGroup:
```
Comment (tigarmo) [inline craft_application/application.py:211]:
```suggestion
    ) -> craft_cli.CommandGroup:
```

**Review by tigarmo — APPROVED** [review]
nice! thanks

**Pull request merged**